app/scripts/migrations/135.ts

```diff
@@ -0,0 +1,289 @@
+import { hasProperty, isObject, NonEmptyArray, Json } from '@metamask/utils';
+import { cloneDeep } from 'lodash';
+
+type CaveatConstraint = {
+  type: string;
+  value: Json;
+};
+
+type PermissionConstraint = {
+  parentCapability: string;
+  caveats: null | NonEmptyArray<CaveatConstraint>;
+};
+
+const PermissionNames = {
+  eth_accounts: 'eth_accounts',
+  permittedChains: 'endowment:permitted-chains',
+};
+
+const BUILT_IN_NETWORKS = {
+  goerli: '0x5',
+  sepolia: '0xaa36a7',
+  mainnet: '0x1',
+  'linea-goerli': '0xe704',
+  'linea-sepolia': '0xe705',
+  'linea-mainnet': '0xe708',
+};
+
+const Caip25CaveatType = 'authorizedScopes';
+const Caip25EndowmentPermissionName = 'endowment:caip25';
+
+const snapsPrefixes = ['npm:', 'local:'] as const;
+
+type VersionedData = {
+  meta: { version: number };
+  data: Record<string, unknown>;
+};
+
+export const version = 135;
+
+/**
+ * This migration transforms `eth_accounts` and `permittedChains` permissions into
+ * an equivalent CAIP-25 permission.
+ *
+ * @param originalVersionedData - Versioned MetaMask extension state, exactly
+ * what we persist to dist.
+ * @param originalVersionedData.meta - State metadata.
+ * @param originalVersionedData.meta.version - The current state version.
+ * @param originalVersionedData.data - The persisted MetaMask state, keyed by
+ * controller.
+ * @returns Updated versioned MetaMask extension state.
+ */
+export async function migrate(
+  originalVersionedData: VersionedData,
+): Promise<VersionedData> {
+  const versionedData = cloneDeep(originalVersionedData);
+  versionedData.meta.version = version;
+  transformState(versionedData.data);
+  return versionedData;
+}
+
+function transformState(state: Record<string, unknown>) {
+  if (
```

Comment:
Not blocking, but I am just noticing that there is a lot of validation code in this migration that I think we could reduce if we make judicious use of Superstruct. I think this would allow us to define a schema and then validate the state against the schema and then get proper types instead of having to typecast everywhere. I'll play around with this and see if I can get something working.

Reply:
I feel like this is a problem with pretty much every single migration that accesses deeply nested state